lib/ansible/config/base.yml

```diff
@@ -1382,6 +1382,18 @@ HOST_PATTERN_MISMATCH:
   - {key: host_pattern_mismatch, section: inventory}
   choices: ['warning', 'error', 'ignore']
   version_added: "2.8"
+TRANSFORM_INVALID_GROUP_CHARS:
+  name: Transform invalid characters in group names
+  default: False
+  description:
+    - Make ansible transform invalid characters in group names supplied by inventory sources.
+    - If 'false' it will allow for the group name but warn about the issue.
```

Comment:
If a customer has any non "-" invalid characters in their inventory groups this won't provide backwards compatibility for the ec2/azure inventory plugins with the scripts, since the scripts would have replaced those non "-" invalid characters with underscores. Tags with invalid characters or IPv6 are possible examples.

The gce script doesn't sanitize, so a blanket toggle like this would keep the gce inventory plugin groups compatible

Reply:
the default is false, but will give users a deprecation notice now, once we toggle to 'true' it will 'autotransform' with a warning for those plugins that don't already do the transform themselves

Reply:
It sounds like you are assuming that some scripts like `ec2.py` sanitize group names. It has a setting for sanitizing group names, yes. But that doesn't mean that it works. I talked about this here:

https://github.com/ansible/ansible/issues/40581#issuecomment-467990027

I can file a bug if you'd like. I have found the behavior to be highly reproducible, and have dived in a little what values it has in code. It appears that `self.replace_dash_in_groups` does take the "True" value (meaning it should sanitize), but I think that either the regex was wrong, or the method `to_safe` was never called.

Reply:
no, assuming the opposite, those using the script will now get a deprecation warning when trying to pass invalid group names

Reply:
@s-hertel I can safely say that in the `ec2.py` script, `to_safe` is not called on the relevant group names, and that the group name `us-east-2` is returned with either `replace_dash_in_groups = True` or `replace_dash_in_groups = False` (with the config loaded correctly, that is ruled out). I can get you more information on this if you want.

Reply:
@AlanCoding I'm not sure if you're understanding me. EC2 script *always* sanitizes, just to varying degrees. Here's my test case. Try adding a bad character in a tag. For instance, host has tags: {'bad:idea': 'here:too'}. Using the script on devel that creates the group `tag_bad_idea_here_too`. Those `:` are replaced by `_` (entirely regardless of replace_dash_in_groups). A filter is being used for bad characters, just not necessarily `-`. So now if you port to using the plugin, you're not going to be able to replicate that behavior while also creating groups with *some* invalid characters (i.e. `-`) because it's all or nothing, no partial sanitization (the group will become `tag_bad:idea:here:too`). Anyway, I don't use it and ripping off the bandaid is probably good. I'm not going to fight for it, just pointing out the deficiency and to have something to point to later if users run into the corner case while trying to replicate the behavior of the script via the plugin. Hopefully no one is doing that but... ¯\_ツ_/¯

Reply:
Related, a global solution like this also poses problems for combining different plugins per playbook run as you may want some plugins to filter and not others.

Reply:
@AlanCoding Ah, btw, replace_dash_in_groups is a [tag specific option](https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.ini#L122). Poorly named, doesn't work on general groups but works great for tags.